### PR TITLE
Reference Space Warp Transformation for ZVZB Forces

### PR DESCRIPTION
**Pull request by rcclay**
## Proposed changes

This implements a preliminary reference version of the space-warp transformation for ZVZB force estimators, given in https://doi.org/10.1063/1.3516208.  Currently, Grad_i(E_L) w.r.t. electron coordinates is handled through finite differences, but replacement with analytic derivatives should reasonably occur within the current infrastructure.  

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Local intel Xeon server.
## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed.  
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCHamiltonians/ACForce.cpp [hunk 1]
```diff
@@ -15,18 +15,23 @@
...
 
 namespace qmcplusplus
 {
 ACForce::ACForce(ParticleSet& source, ParticleSet& target, TrialWaveFunction& psi_in, QMCHamiltonian& H)
-    : ions(source), elns(target), psi(psi_in), ham(H), FirstForceIndex(-1), Nions(0)
+    : ions(source), elns(target), psi(psi_in), ham(H), FirstForceIndex(-1), Nions(0), useSpaceWarp(false), swt(target,source)
 {
   prefix = "ACForce";
   myName = prefix;
   Nions  = ions.getTotalNum();
```
Comment (ye-luo) [inline src/QMCHamiltonians/ACForce.cpp:27]: please make Nions const in its definition and assign its value at the constructor initializer list.

File: src/QMCHamiltonians/ACForce.cpp [hunk 2]
```diff
@@ -52,10 +78,20 @@ ACForce::Return_t ACForce::evaluate(ParticleSet& P)
...
+  Force_t el_grad;
+  el_grad.resize(P.getTotalNum());
+  el_grad=0;
   //This function returns d/dR of the sum of all observables in the physical hamiltonian.
   //Note that the sign will be flipped based on definition of force = -d/dR.
   Value = ham.evaluateIonDerivs(P, ions, psi, hf_force, pulay_force, wf_grad);
+  
+  if(useSpaceWarp)
+  {
+    ham.evaluateElecGrad(P,psi,el_grad,delta);
```
Comment (ye-luo) [inline src/QMCHamiltonians/ACForce.cpp:92]: move the el_grad inside if(useSpaceWarp)

File: src/QMCHamiltonians/ACForce.cpp [hunk 3]
```diff
@@ -75,10 +111,25 @@ void ACForce::addObservables(PropertySetType& plist, BufferType& collectables)
       pulayname << prefix << "_pulay_" << iat << "_" << x;
       wfgradname1 << prefix << "_Ewfgrad_" << iat << "_" << x;
       wfgradname2 << prefix << "_wfgrad_" << iat << "_" << x;
+
       plist.add(hfname.str());
       plist.add(pulayname.str());
       plist.add(wfgradname1.str());
       plist.add(wfgradname2.str());
+
+      //Remove when ACForce is production ready.
```
Comment (ye-luo) [inline src/QMCHamiltonians/ACForce.cpp:120]: Add `TODO: ` before Remove.

File: src/QMCHamiltonians/ACForce.h [hunk 4]
```diff
@@ -54,6 +55,8 @@ struct ACForce : public OperatorBase
   /** Evaluate **/
   Return_t evaluate(ParticleSet& P);
 
+  RealType delta; //finite difference time step
```
Comment (ye-luo) [inline src/QMCHamiltonians/ACForce.h:58]: Please use Doxygen style comment

File: src/QMCHamiltonians/OperatorBase.h [hunk 5]
```diff
@@ -246,6 +246,7 @@ struct OperatorBase : public QMCTraits
    *@return the value of the Hamiltonian component
    */
   virtual Return_t evaluate(ParticleSet& P) = 0;
+  virtual Return_t evaluateDeterministic(ParticleSet& P);
```
Comment (ye-luo) [inline src/QMCHamiltonians/OperatorBase.h:249]: Documentation is needed for this function. Please describe the desired behaviour.

File: src/QMCHamiltonians/OperatorBase.cpp [hunk 6]
```diff
@@ -35,6 +35,10 @@ OperatorBase::OperatorBase() : myIndex(-1), Dependants(0), Value(0.0), tWalker(0
   UpdateMode.set(PRIMARY, 1);
 }
 
+OperatorBase::Return_t OperatorBase::evaluateDeterministic(ParticleSet& P)
+{
+  return evaluate(P);
```
Comment (ye-luo) [inline src/QMCHamiltonians/OperatorBase.cpp:40]: Is this a safe fallback? The fallback seems non-deterministic.

File: src/QMCHamiltonians/SpaceWarpTransformation.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+{
+  typedef ParticleSet::ParticleScalar_t ParticleScalar_t;
+  typedef ParticleSet::ParticlePos_t Force_t;
+  typedef ParticleSet::ParticleGradient_t ParticleGradient_t;
+
+  SpaceWarpTransformation(ParticleSet& elns, ParticleSet& ions);
+  ~SpaceWarpTransformation(){};
+
+ 
+  RealType f(RealType r);  //space warp transformation function F.
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.h:36]: please use doxygen style comments for functions.

File: src/QMCHamiltonians/SpaceWarpTransformation.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                                                      //space warp components and gradients.
+
+  void getSWT(int iat, ParticleScalar_t& w, Force_t& grad_w); //For each ion component iat, this generates all the required space warp quantities to 
+                                                              //generate the "space warp" contribution to the iat-th force component.
+                                                              
+  void computeSWT(ParticleSet& elec, ParticleSet& ions, Force_t& dEl, ParticleGradient_t& dlogpsi, Force_t& el_contribution, Force_t& psi_contribution); //Takes in precomputed grad(E_L) and grad(logPsi), and uses
+                                                                                                                        //this to compute the ZV (el_contribution) and ZB (psi_contribution) 
+                                                                                                                        //space warp contributions.                                                               
+
+  int myTableIndex;
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.h:51]: const myTableIndex, Nelec, Nions

File: src/QMCHamiltonians/SpaceWarpTransformation.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
+#include "QMCHamiltonians/SpaceWarpTransformation.h"
+#include "Particle/DistanceTableData.h"
+#include "type_traits/scalar_traits.h"
+namespace qmcplusplus
+{
+
+SpaceWarpTransformation::SpaceWarpTransformation(ParticleSet& elns, ParticleSet& ions):
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.cpp:7]: `const ParticleSet& ions`

File: src/QMCHamiltonians/ACForce.cpp [hunk 10]
```diff
@@ -15,18 +15,23 @@
  */
 #include "ACForce.h"
 #include <sstream>
+#include "OhmmsData/AttributeSet.h"
 
 namespace qmcplusplus
 {
 ACForce::ACForce(ParticleSet& source, ParticleSet& target, TrialWaveFunction& psi_in, QMCHamiltonian& H)
-    : ions(source), elns(target), psi(psi_in), ham(H), FirstForceIndex(-1), Nions(0)
+    : ions(source), elns(target), psi(psi_in), ham(H), FirstForceIndex(-1), Nions(0), useSpaceWarp(false), swt(target,source)
```
Comment (ye-luo) [inline src/QMCHamiltonians/ACForce.cpp:23]: Does ACForce modify ions? If no, it is better to set `ions` and `source` as const reference.

File: src/QMCHamiltonians/SpaceWarpTransformation.h [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+#include "Particle/ParticleSet.h"
+
+namespace qmcplusplus
+{
+/** @ingroup hamiltonian
+ *\brief This implements the differential space warp transformation for ZVZB estimators given by Sorella & Capriotti
+ *       J. Chem. Phys., 133, 23411 (2010), https://doi.org/10.1063/1.3516208 
+ *
+ */
+struct SpaceWarpTransformation : public QMCTraits
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.h:26]: please use class instead of struct and put proper public, private.

**Review by ye-luo — CHANGES_REQUESTED** [review]
Please clang-format the modified source code will also be helpful.

File: src/QMCHamiltonians/OperatorBase.cpp (hunk 6 above)
Comment (rcclay) [reply src/QMCHamiltonians/OperatorBase.cpp:40]: Unfortunately, the correct behavior of this routine requires estimators with nondeterministic components in their evaluate() have this method overloaded.  Provided that is done, it's safe.  Is this what you were referring to?

File: src/QMCHamiltonians/ACForce.cpp (hunk 10 above)
Comment (rcclay) [reply src/QMCHamiltonians/ACForce.cpp:23]: ACForce potentially modifies ions, if QMCHamiltonian.evaluateIonDerivs starts doing finite differences or something.  Would be complicated to make this const here.  For SpaceWarpTransformation, there's no good reason the ions should be modified, so I'll make those consts.

**Comment by rcclay** [comment]
@ye-luo  OK.  Short of the making ions "const" issue and the default behavior of evaluateDeterministic, I believe I have addressed all your concerns.

File: src/QMCHamiltonians/OperatorBase.cpp (hunk 6 above)
Comment (ye-luo) [reply src/QMCHamiltonians/OperatorBase.cpp:40]: "the correct behavior of this routine requires estimators with nondeterministic components in their evaluate() have this method overloaded" this piece of info should be recorded in the OperatorBase.h before the function.

File: src/QMCHamiltonians/OperatorBase.h (hunk 5 above)
Comment (ye-luo) [reply src/QMCHamiltonians/OperatorBase.h:253]: Could you say a bit more about waht "deterministically" means?

File: src/QMCHamiltonians/NonLocalECPotential.h [hunk 12]
```diff
@@ -152,7 +152,7 @@ class NonLocalECPotential : public OperatorBase, public ForceBase
    * @param P particle set
    * @param Tmove whether Txy for Tmove is updated
    */
-  void evaluateImpl(ParticleSet& P, bool Tmove);
+  void evaluateImpl(ParticleSet& P, bool Tmove, bool keepGrid=false);
```
Comment (ye-luo) [inline src/QMCHamiltonians/NonLocalECPotential.h:155]: please document keepGrid. This file seems not being formatted.

File: src/QMCHamiltonians/QMCHamiltonian.h [hunk 13]
```diff
@@ -231,6 +231,12 @@ class QMCHamiltonian
    */
   FullPrecRealType evaluate(ParticleSet& P);
 
+  /** evaluate Local Energy deterministically.  Defaults to evaluate(P) for operators without a stochastic component. For the nonlocal PP, the quadrature grid is not rerandomized.  
+   * @param P ParticleSet
+   * @return Local energy. 
+   *
```
Comment (ye-luo) [inline src/QMCHamiltonians/QMCHamiltonian.h:237]: Could you remove this additional empty line.

File: src/QMCHamiltonians/SpaceWarpTransformation.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+class SpaceWarpTransformation : public QMCTraits
+{
+  typedef ParticleSet::ParticleScalar_t ParticleScalar_t;
+  typedef ParticleSet::ParticlePos_t Force_t;
+  typedef ParticleSet::ParticleGradient_t ParticleGradient_t;
+
+  public:
+
+  SpaceWarpTransformation(ParticleSet& elns, const ParticleSet& ions);
+  ~SpaceWarpTransformation(){};
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.h:35]: Please remove this empty destructor line.

File: src/QMCHamiltonians/SpaceWarpTransformation.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+  RealType f(RealType r);  ///< Space warp transformation function F.  Argument is distance.
+  RealType df(RealType r); ///< The derivative of F w.r.t. r.
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.h:38]: Please use doxygen style.

File: src/QMCHamiltonians/SpaceWarpTransformation.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+   *
+   * \param[in] P, the electron particle set.  
+   * \param[in] ions, the ion particle set.  
+   */
+  void computeSWTIntermediates(ParticleSet& P,
+                               const ParticleSet& ions); 
+                                                         
+
+
+  const int myTableIndex;
```
Comment (ye-luo) [inline src/QMCHamiltonians/SpaceWarpTransformation.h:90]: please comment what table it is? My guess is electron ion table.

**Review by ye-luo — COMMENTED** [review]
Thank you for helping with the code quality improvements. Only a few minor issues left.

File: src/QMCHamiltonians/ACForce.cpp (hunk 10 above)
Comment (ye-luo) [reply src/QMCHamiltonians/ACForce.cpp]: That makes sense.

**Review by ye-luo — APPROVED** [review]

**Pull request merged**